### DISCUSSION TITLE
Applicability of PPO to algo-trading and general questions on this topic

### DISCUSSION THREAD

**Discussion by artiko88** [theOGognf/finagg#58] [General]
Hey!

First off, thank you for making this library available to the public. Really like the philosophy of the library.

I have to say that my current experiments are using SB3, since rl8 is more flexible I would gladly switch to it, provided its flexibility may help solve the problem i am facing.

I am new to RL and decided to try in order to solve a task in the field of algo-trading crypto assets. The family of strategies that I am developing and deploy is market making and I am using an extension of classic Avellaneda-Stoikov based model. Basically, it assesses intensity of market orders arrival (how far from midprice they arrive) and volatility as an STD of price difference within a time window and, using sophisticated formula, derives how deep from the midprice to send orders and how much to skew those orders around the midprice to manage its inventory. Strategy "ticks" every 100ms and sends both buy and sell orders with the intention of capturing spread between them. At each 100ms it cancels orders that were not filled and send a new portion (if recalculated price is large enough, that is). When some of the sent order is hit, bid and ask prices are skewed, so that side that has inventory is now closer to the midprice and thus have a higher probability of being filled. Every 5 seconds it reassesses the state by recalculating volatility and arrival rates and decides upon depth and skew values.

The strategy has shown good results both in backtest and in live application, only although it manages risk to some extent, it still incurs large losses on large swings of price that happen in crypto world very often. So in order to deal with the risk and maybe improve results, I built a gymnasium compatible environment based on an event-driven backtester and started experimenting. The idea is to use reinforcement learning to decide upon depth and skew value and instead of using arrival rate and volatility, use a wide range of market defining features along with agent defining features. So I set it up like this: agent observes the state, decides upon spread and skew value, passes it to backtester which uses those values for 5 seconds, collects a reward value (difference in PNL with some additional rewards and penalties for managing risk etc) and provides another set of spread and skew values. This way training is faster, since I skip providing prices for each order being sent and instead send params to a strategy that use those locked actions for some time. features that i use are mostly based on microstructure (LOB), correlated assets and current trade activity. Features are generated in a massive amount and then only important features that are filtered by another library are left. The importance is detected by using a classification model which predicts whether next period of prices will differ in terms of std by more than X than current period. Currently i use about 50-60 features in the state. The period is 5 seconds - amount of time that provided params are used by the strategy. I use continuous actions bound between -1 and 1 which are then de-normalized in the backtester to a necessary range. I try to train the agent on 10 days worth of data which amounts to about 345_600 unique observations. In order to avoid "remembering" the environment, I split the data in hours and feed those hours randomly.

The problem with this all is that the agent does not learn a policy that allows it to show results at least on par with analytical strategy (Avellaneda based), not to say a better one. It eventually chooses a very thin fraction of action space for both actions and use those on almost every step. Action are discretized in the backtester, so it basically uses mostly the same action most of the time. 

What is interesting is that sometimes loss functions decrease largely. And explained variance also may grow some and then decrease.

What I have tried already to fix it:
- use VecNormalize for rewards and observation standardization
- reward function engineering: using VecNormnalize, normalization by an average reward from analytical strategy, using return ratio vs difference, rewarding action diversity (e.g. providing more reward when there more than X unique action in recent steps or subtracting some value when there are less
- hyperparameter tuning: clip fraction between 0.05 to 0.5, learning rate from high to low, ent_coeff from 0 to 0.05, learning rate schedule. Larger batch sizes, n_steps and epochs.
- Different algo - I tried RecurrentPPO, DroQ, SAC. Most of experiments are with PPO, though.
- Leaving the environment to run for 30-40M steps. 

So my questions are:
- since this environment is far from being fully observable, can regular PPO be used to find optimal actions for this task?
- given the nature of environment, should one use a higher learning rate or a lower one? Should higher ent_coeff be used? How about clip_range? 
- Should a more sophisticated or larger neural network be used instead of default [64,64] MlpPolicy used in SB3?
- Should some other layers be used on top of MlpPolicy? LSTM, Attention?

This may look like I want somebody to spoon-feed me and maybe it is in fact what i want, lol given current state of desperation i am in after so many weeks of experiments :) But I would really appreciate any pointers that might be helpful.

Thanks!

**Comment by theOGognf** [comment]
Cool strategy and use case!

> I have to say that my current experiments are using SB3, since rl8 is more flexible I would gladly switch to it, provided its flexibility may help solve the problem i am facing.

rl8 is only good for cases where the environment and policy can be implemented and ran using the same GPU. Sounds like your environment would be tough to vectorize on a GPU because of the backtesting dependency

> Currently i use about 50-60 features in the state

RL is good for sequential decision-making where a policy's action affects its state. I can't say I fully understand your formulation, but if the state isn't affected by the policy's action, then it makes sense that the policy converges to something similar to the Avellaneda-based solution. E.g., a use case that only used stock prices as the observation wouldn't have its observation be affected by buy/sell actions. E.g., a use case that was on portfolio optimization whose observation was percent of portfolio in bonds (or some other asset) would see its observation affected by the action

> Action are discretized in the backtester, so it basically uses mostly the same action most of the time.

Why not discretize the action space for the policy directly then instead of using a continuous action space?

> What is interesting is that sometimes loss functions decrease largely. And explained variance also may grow some and then decrease.

Loss is hard to interpret for RL

> What I have tried already to fix it:
> 
>     * use VecNormalize for rewards and observation standardization
> 
>     * reward function engineering: using VecNormnalize, normalization by an average reward from analytical strategy, using return ratio vs difference, rewarding action diversity (e.g. providing more reward when there more than X unique action in recent steps or subtracting some value when there are less
> 
>     * hyperparameter tuning: clip fraction between 0.05 to 0.5, learning rate from high to low, ent_coeff from 0 to 0.05, learning rate schedule. Larger batch sizes, n_steps and epochs.
> 
>     * Different algo - I tried RecurrentPPO, DroQ, SAC. Most of experiments are with PPO, though.
> 
>     * Leaving the environment to run for 30-40M steps.

Good that you tried these things

> So my questions are:
> 
>     * since this environment is far from being fully observable, can regular PPO be used to find optimal actions for this task?
> 
>     * given the nature of environment, should one use a higher learning rate or a lower one? Should higher ent_coeff be used? How about clip_range?
> 
>     * Should a more sophisticated or larger neural network be used instead of default [64,64] MlpPolicy used in SB3?
> 
>     * Should some other layers be used on top of MlpPolicy? LSTM, Attention?

Good questions. No one knows. Best you can do is try them and see

Overall, sounds like youre struggling with the classic risk vs reward tradeoff. Make sure your reward incorporates risk in its formulation if you want to minimize risk explicitly. That way you don't have to depend on the policy to find a less risky solution as randomly. Make sure you're properly comparing the policy's performance to the Avellaneda-based method as well (measure risk and reward for both across equivalent time frames)

My only suggestion would be to instead of making the policy determine depth and skew values directly, have it determine depth and skew value perturbations from the Avellaneda-based method. That is, use the Avellaneda-based method as part of your strategy, but have the policy make small updates to it. That way its easier to learn something that might outperform it since its effectively using it as a baseline

**Reply by artiko88** [reply]
> rl8 is only good for cases where the environment and policy can be implemented and ran using the same GPU. Sounds like your environment would be tough to vectorize on a GPU because of the backtesting dependency

Didn't realize that rl8 is all GPU. Bummer :(



> RL is good for sequential decision-making where a policy's action affects its state. I can't say I fully understand your formulation, but if the state isn't affected by the policy's action, then it makes sense that the policy converges to something similar to the Avellaneda-based solution. E.g., a use case that only used stock prices as the observation wouldn't have its observation be affected by buy/sell actions. E.g., a use case that was on portfolio optimization whose observation was percent of portfolio in bonds (or some other asset) would see its observation affected by the action

Well, some part of the state is affected by the action of the agent. There is inventory - size of the position that - which can be either positive (long) or negative (short). It varies, but i somewhat settled in the value of 25 order sizes as maximum possible position, above (or below) which strategy would refuse to increase it. This value in the state is normalized between -1 and 1. Then there is fill rate - how many orders are filled within last rolling window size. Fill rate has a sign so that we know if orders are filled on some particular side (long/short). Good to know so that if fill rate increases and we lose because of it, agent can have the depth corrected. There are a few more.

And the bad part is that the policy in fact does not converge even near the performance of Avellaneda based strategy. What it does, it converges on a pair of depth/skew values which it varies a tiny bit between steps, but generally, given the discretization of the action space, for backtester these are same values. So instead of being an agent that reacts to the state by changing depth/skew, it simply uses same actions over and over.

As for why not discretize... Well, initial thought process was it could be easier for the agent deal with a continuous values rather than discrete values. But this is what I would like to in practice.

>Overall, sounds like you're struggling with the classic risk vs reward tradeoff. Make sure your reward incorporates risk in its formulation if you want to minimize risk explicitly. That way you don't have to depend on the policy to find a less risky solution as randomly. Make sure you're properly comparing the policy's performance to the Avellaneda-based method as well (measure risk and reward for both across equivalent time frames)

Risk is somewhat incorporated since I am comparing ewma's of PNL values between two steps. So each step skips 5seconds of data and PNL value is sampled at 100ms intervals. By comparing mean values I incorporate drawdowns as well and by using EWMA i put more meaning on the last values of those steps.

As for providing adjustments for AS based strategy - this is what i have tried. AS has a set of hyperparameters like gamma and delta which basically deal with how deep to place orders from the mid price and how aggressively to skew. The idea (which i copied from one of the papers on the topic) was to provide optimal parameters to an AS based strategy. Initial parameters were set with values centered around values that were found optimal for this particular historical period with Optuna using AS strategy (let's find optimal AS params for a historical period, let's use them as a basis and allow RL agent to choose from 0.5*optimal_param and 2*optimal_param). This also failed. 

>Good questions. No one knows. Best you can do is try them and see

Yeah, probably this is the only way it can be done. Feels like I am missing something very fundamental. Like, rewards are absolutely not informative. Or they are, but sometimes large values of rewards put the model out of whack and even clip_range param does not help with it (although this is unlikely, since I tried clipping rewards). So the policy finds a single combination of depth/skew as "optimal". But things like volatility, trade volume, dependency on correlated assets, intensity of trade arrivals - they all change from time to time and the agent should learn to use optimal set of depth/skew combinations for different representations of the state, which it refuses to do and uses something universal :(

**Reply by theOGognf** [reply]
> As for why not discretize... Well, initial thought process was it could be easier for the agent deal with a continuous values rather than discrete values. But this is what I would like to in practice.

Definitely recommend doing this

> As for providing adjustments for AS based strategy - this is what i have tried. AS has a set of hyperparameters like gamma and delta which basically deal with how deep to place orders from the mid price and how aggressively to skew. The idea (which i copied from one of the papers on the topic) was to provide optimal parameters to an AS based strategy. Initial parameters were set with values centered around values that were found optimal for this particular historical period with Optuna using AS strategy (let's find optimal AS params 

This sounds similar to what I'm suggesting, but what I mean is to set gamma and delta fixed and have the RL agent choose from discrete options that're perturbations from gamma and delta. So the final would be [(gamma + d_gamma),  (delta + d_delta)] where d_gamma and d_delta is the policy's action

Yeah RL can be a bummer sometimes. Best of luck

**Reply by artiko88** [reply]
Will def try those suggestions. Thanks!
